libs/wingsdk/src/std/map.ts

```diff
@@ -6,6 +6,16 @@ import { Array } from "./array";
 import { T1 } from "./generics";
 import { InflightClient } from "../core";
 
+/**
+ * Array entry representation
+ */
+export interface ArrayEntry {
+  /** The entry key */
+  readonly key: string;
+  /** The entry value */
+  readonly value: any;
```

Comment:
Hey @Warkanlock thanks so much for the detailed description and deep diving research to try and find something that works. 

I agree with you that given the current limitations that we cannot implement this feature without the use of `any` as the value type. However, I think since there exists a less convenient work around at the moment (iterate on `keys()` and get type safe values that way) that we should hold off for the moment until we can implement this in a typesafe way.

That said I want to allow others to chime in here as well. - thoughts? @Chriscbr, @MarkMcCulloh, @eladb, @yoav-steinberg

Reply:
I think I tend to agree the the API feels like it's less useful if all the entries you receive are typed as `any`. I'd be in favor of documenting a workaround for use cases where you want to iterate over keys and values together.

Reply:
I agree with both; also, we need to keep in mind that JSII does not allow generics in this specific scenario, so I proposed the following:

1. Remove the `.key` and `.value` API and stick with the Javascript native signature. [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries)
2. This will allow us to keep the value at least as it is without the need to type it

**_However_**, playing around with the `Map` type, I realized there's no way to init a map other than the type of inception. At least I'm trying to init a `Map` from the playground (latest version available) 

E.g.:
```wing
let map = { "foo" => 1, "bar" => "test" };
                                                      ^^^ Expected type to be "num", but I got "str" instead
```

Having said that, we can "assume" that the type of the `Map` to be initialized and returned should have the same value type as the first element of the `Map`. From that we can ensure type-safety at cost of having only one type per map (not modified in the scope of this pull-request though)

Let me know if I'm losing some context in the middle since it's my first PR to the repo, and I might be confused by now

Reply:
More information on this issue,

Typescript, by default, does not guarantee `type safety' for each element, even on the native `Object.entries()`. [The guarantee the type from a list of types](https://github.com/microsoft/TypeScript/blob/v4.6.2/lib/lib.es2017.object.d.ts#L34-L44). 

So, technically, in case we want to keep as close to the native javascript interface as possible (and to maintain the behaviour of entries), I don't think we will be able to do that without some trade-off in the middle.

An example:
```ts
const elements = {
    "1": "Hydrogen",
    "2": "Helium",
    "3": 12,
} 
```

You will get a type as:

```ts
signature(...ArrayLike<string | number>): [string, string | number][]
```

Which is technically the behavior indicated in the snippet I submitted above. However, it's virtually impossible to implement generics using JSII (perhaps some typing magic will do the trick, but the compilation process will fail as soon as a `Generic` is introduced).

I'm glad to hear your opinions on this since I'm new to the codebase, although, at this point, I think it's worth asking a question: Is it better not to have the function at all than to have it and not have it with the guarantee of type safety?

Reply:
@hasanaburayyan I know that this problem is not that serious, although it would be nice to see if we will continue to find a possible solution or if we will have to make trade-offs to put it in place.

Do you have any thoughts on this?

Reply:
@Warkanlock sorry I missed the notification from your message 😅

To your question:

> Is it better not to have the function at all than to have it and not have it with the guarantee of type safety?

In my personal opinion (fwiw), its better to not allow the function. Especially since there is a work around for achieving iterating through a map. Type safety is pretty fundamental to Wing and if we allow users to break the type safety it needs to be through a "Opt-In" path. Having a standard type expose a function that breaks typesafety, sounds like it may set users up for a pretty complicated debug experience.

That said, its been a little while since I have last looked into this issue of enhancing Wing Generics, right now we achieve some basic generic functionality by using a marker class called `T1` then the compiler does a hydration process replacing these types. Though it seems we are getting close to really needing to dive into a more sophisticated solution. 

> Having said that, we can "assume" that the type of the Map to be initialized and returned should have the same value type as the first element of the Map. From that, we can ensure type safety at cost of having only one type per map (not modified in the scope of this pull request though)

So yes the observation is correct, and intentional. In fact all Container types in Wing are homogenous by design. The only type that differs from this rule is Json which can be heterogenous.